examples/cis1-wgtu/src/lib.rs

```diff
@@ -499,7 +497,61 @@ fn contract_balance_of<A: HasActions>(
         response.push((query, amount));
     }
     // Send back the response.
-    Ok(send(&sender, callback.as_ref(), Amount::zero(), &BalanceOfQueryResponse::from(response)))
+    Ok(send(
+        &result_contract,
+        result_hook.as_ref(),
+        Amount::zero(),
+        &BalanceOfQueryResponse::from(response),
+    ))
+}
+
+#[allow(dead_code)]
+type ContractTokenMetadataQueryParams = TokenMetadataQueryParams<ContractTokenId>;
+
+/// Get the token metadata URLs and checksums given a list of token IDs and it
+/// takes a contract address plus contract function to invoke with the result.
+///
+/// It rejects if:
+/// - It fails to parse the parameter.
+/// - Any of the queried `token_id` does not exist.
+/// - Message sent back with the result rejects.
+#[receive(
+    contract = "CIS1-wGTU",
+    name = "tokenMetadata",
+    parameter = "ContractTokenMetadataQueryParams"
+)]
+fn contract_token_metadata<A: HasActions>(
+    ctx: &impl HasReceiveContext,
+    _state: &mut State,
+) -> ContractResult<A> {
+    let mut cursor = ctx.parameter_cursor();
+    // Parse the contract address to receive the result.
+    let result_contract: ContractAddress = cursor.get()?;
+    // Parse the contract function name to call with the result.
+    let result_hook: OwnedReceiveName = cursor.get()?;
+    // Parse the number of queries.
+    let queries_length: u8 = cursor.get()?;
+
+    // Build the response.
+    let mut response = Vec::with_capacity(queries_length.into());
+    for _ in 0..queries_length {
+        let token_id: ContractTokenId = cursor.get()?;
```

Comment:
This technically means that there can be leftover data in the parameter which might or might not be an issue.

You could potentially check at the end of the loop that you have reached the end of the parameter.

Reply:
I have made an issue #64 which will provide the api necessary for this.